### PR TITLE
Allow the building of dqlite for develop

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following allows the building of juju with a dqlite instance. This is very
rough around the edges, as it doesn't actually do anything as nobody is
calling the dbaccessor internally. Accessing the REPL downloads socat
and rlwrap (read line wrap) in a container to prevent breaking of the local
system setup.

The makefile has been hacked together and probably requires updating to
the latest makefile format. Might be worth getting @tlm to fix the makefile after
it lands.

## Installing Juju

```sh
$ make go-install
```

## Bootstrapping to LXD

```sh
$ juju bootstrap lxd test --build-agent
```

## Accessing the REPL

This only works on LXD for now and we don't have any databases in this PR, so `.open` doesn't work. In the `dqlite-scaffolding` branch we had the logs db, so we could `.open` the database there.

```sh
$ make juju-dqlite-repl
```

## Data model

Included is a CLI tool that analyzes the set of structs defined in Juju's `state` package and attempts to generate an ER-like diagram to aid with the data modeling work which is part of the transition from mongo to sqlite. 

The actual diagram looks more like a UML class diagram as the number of entities (tables) and attributes makes it more difficult for graphviz to figure out a suitable layout for the graph nodes.

The tool automatically normalizes both table and field names to the underscore-cased format that folks are more familiar with when working with SQL databases. In addition, the tool will attempt to cluster together tables that share a common name prefix.

Future work ideas/suggestions:
- Infer relations between tables by analyzing the names of fields that serve as foreign keys (this can also yield a more accurate clustering result). Alternatively, the tool could be seeded with an external (JSON/YAML etc.) file to provide hints as to which tables should be grouped together.
- The tool currently lists the Go type for each attribute. However, for attributes that are nested objects or collections of objects we could automatically generate auxiliary tables for storing them and also annotate the graph with edges that will help identify the relationships between fields.

Generating the diagram

```sh
$ cd scripts/dqlite/cmd

# NOTE 1: use 'open' instead of xdg-open if running on OSX
# NOTE 2: you must use `fdp` as your layout engine to get nice, rectangular output for the graph!
$ go run infer_schema.go | fdp -o erd.svg -Tsvg && xdg-open erd.svg
```

Yields something like this:

erd

### CODE REVIEW

File: worker/dbaccessor/repl.go [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+	"math/rand"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v2"
```
Comment (manadart) [inline worker/dbaccessor/repl.go:20]: Use v3. Also need a `go mod tidy`.

**Review by manadart — CHANGES_REQUESTED** [review]
Couple of things to tidy up.

Can you also elaborate in the description regarding how it hangs together? Binaries in S3 etc.

**Review by manadart — APPROVED** [review]
Thanks for this.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**